core/src/main/java/net/opentsdb/storage/schemas/tsdb1x/Schema.java

```diff
@@ -635,22 +646,72 @@ public int tagvWidth() {
    * @param row_key A non-null and non-empty row key.
    */
   public void prefixKeyWithSalt(final byte[] row_key) {
-    // TODO - implement
+    if (salt_width < 1) {
+      return;
+    }
+    if ((row_key.length < salt_width + metric_width) ||    
+        (Bytes.memcmp(row_key, new byte[salt_width + metric_width], 
+            salt_width, metric_width) == 0)) {    
+        //Metric id is 0, which means it is a global row. Don't prefix it with salt   
+        return;
+      }
+    
+    if (timeless_salting) {
+      // we want the metric and tags, not the timestamp
+      int hash = 1;
+      for (int i = salt_width; i < row_key.length; i++) {
+        hash = 31 * hash + row_key[i];
+        if (i + 1 == salt_width + metric_width) {
+          i = salt_width + metric_width + Const.TIMESTAMP_BYTES - 1;
+        }
+      }
+      int modulo = hash % salt_buckets;
+      if (modulo < 0) {
+        // make sure we return a positive salt.
+        modulo = modulo * -1;
+      }
+      prefixKeyWithSalt(row_key, modulo);
+    } else if (old_salting) {
+      // DON'T DO THIS! Don't USE IT!
+      int modulo = (new String(Arrays.copyOfRange(row_key, salt_width,    
+          row_key.length), Const.ASCII_US_CHARSET)).hashCode() % salt_buckets;   
+      if (modulo < 0) {
+        // make sure we return a positive salt.
+        modulo = modulo * -1;   
+      }
+      prefixKeyWithSalt(row_key, modulo);
+    } else {
+      int hash = 1;
+      for (int i = salt_width; i < row_key.length; i++) {
+        hash = 31 * hash + row_key[i];
+      }
+      int modulo = hash % salt_buckets;
+      if (modulo < 0) {
+        // make sure we return a positive salt.
+        modulo = modulo * -1;
+      }
+      prefixKeyWithSalt(row_key, modulo);
+    }
   }
   
+  /**
+   * Sets the bucket on the row key. The row key must be pre-allocated
+   * the proper number of salt bytes at the start of the array.
+   * @param row_key A non-null and non-empty row key.
+   * @param bucket A 0 based positive bucket ID.
+   */
   public void prefixKeyWithSalt(final byte[] row_key, final int bucket) {
     if (salt_width == 1) {
       row_key[0] = (byte) bucket;
       return;
     }
     
-    final byte[] salt = new byte[salt_width];
     int shift = 0;
     for (int i = 0; i <= salt_width; i++) {
-      salt[salt_width - i] = (byte) (bucket >>> shift);
+      row_key[salt_width - i] = 0;
```

Comment:
This will replace the first byte of the metric name when i = 0.
Also, total iterations are salt_width + 1. Starting from i = 1 may fix it.